server/src/main/java/org/opensearch/extensions/action/ExtensionActions.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.extensions.action;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.extensions.DiscoveryExtension;
+import org.opensearch.extensions.ExtensionBooleanResponse;
+import org.opensearch.extensions.ExtensionsOrchestrator;
+import org.opensearch.extensions.RegisterTransportActionsRequest;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.ActionNotFoundTransportException;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class manages TransportActions for extensions
+ *
+ * @opensearch.internal
+ */
+public class ExtensionActions {
+    private static final Logger logger = LogManager.getLogger(ExtensionActions.class);
+    private Map<String, DiscoveryExtension> actionsMap;
+    private final Map<String, DiscoveryExtension> extensionIdMap;
+    private final TransportService transportService;
+    private final NodeClient client;
+
+    public ExtensionActions(Map<String, DiscoveryExtension> extensionIdMap, TransportService transportService, NodeClient client) {
+        this.actionsMap = new HashMap<>();
+        this.extensionIdMap = extensionIdMap;
+        this.transportService = transportService;
+        this.client = client;
+    }
+
+    void registerAction(String action, DiscoveryExtension extension) throws IllegalArgumentException {
+        if (actionsMap.containsKey(action)) {
+            throw new IllegalArgumentException("The " + action + " you are trying to register is already registered");
+        }
+        actionsMap.putIfAbsent(action, extension);
+    }
+
+    public DiscoveryExtension getExtension(String action) {
+        return actionsMap.get(action);
+    }
+
+    /**
+     * Handles a {@link RegisterTransportActionsRequest}.
+     *
+     * @param transportActionsRequest  The request to handle.
+     * @return  A {@link ExtensionBooleanResponse} indicating success.
+     */
+    public TransportResponse handleRegisterTransportActionsRequest(RegisterTransportActionsRequest transportActionsRequest) {
```

Comment:
Separate handler class here?

Reply:
Its just one single method, I'd prefer to have all the functionality with the same class as it's taking care of transport actions. If the handlers grow to multiple of them we can start splitting them.